tt_eager/tt_dnn/op_library/run_operation.cpp

```diff
@@ -99,45 +122,50 @@ std::vector<Tensor> run_host_operation(const HostOperation& operation, const std
-    const std::vector<std::optional<Tensor>>& optional_output_tensors) {
+    const DeviceOperation<OutputTensors>& operation,
+    const Tensors& input_tensors,
+    const OptionalConstTensors& optional_input_tensors,
+    const OptionalTensors& optional_output_tensors) {
     ZoneScopedN("TT_DNN_DEVICE_OP");
     uint32_t op_id = assign_id();
 
     std::function<std::variant<std::shared_ptr<Program>, std::reference_wrapper<Program>>(
-        const DeviceOperation&,
-        const std::vector<Tensor>&,
-        const std::vector<std::optional<const Tensor>>&,
-        std::vector<Tensor>&)>
+        const DeviceOperation<OutputTensors>&,
+        const Tensors&,
+        const OptionalConstTensors&,
+        OutputTensors&)>
         get_or_create_program;
     auto& program_cache = input_tensors[0].device()->program_cache;
     if (program_cache.is_enabled()) {
-        get_or_create_program = [&program_cache](const DeviceOperation& operation,
-                                   const std::vector<Tensor>& input_tensors,
-                                   const std::vector<std::optional<const Tensor>>& optional_input_tensors,
-                                   std::vector<Tensor>& output_tensors) -> std::reference_wrapper<Program> {
+        get_or_create_program = [&program_cache](const DeviceOperation<OutputTensors>& operation,
+                                   const Tensors& input_tensors,
+                                   const OptionalConstTensors& optional_input_tensors,
+                                   OutputTensors& output_tensors) -> std::reference_wrapper<Program> {
 
             auto program_hash = operation.compute_program_hash(input_tensors, optional_input_tensors);
             auto program_ptr = program_cache.find(program_hash);
 
             bool cache_hit = program_ptr.has_value();
             if (not cache_hit) {
-                program_ptr = std::make_shared<operation::ProgramWithCallbacks>(operation.create_program(input_tensors, optional_input_tensors, output_tensors));
+                program_ptr = std::make_shared<operation::CacheableProgram<OutputTensors>>(operation.create_program(input_tensors, optional_input_tensors, output_tensors));
                 program_cache.insert(program_hash, program_ptr.value());
             }
-            auto& program_with_callbacks = *(reinterpret_cast<operation::ProgramWithCallbacks*>(program_ptr.value().get()));
+            auto& program_with_callbacks = *(reinterpret_cast<operation::CacheableProgram<OutputTensors>*>(program_ptr.value().get()));
             TT_ASSERT(program_with_callbacks.supports_program_cache());
 
             if (cache_hit) {
                 ZoneScopedN("Cache_hit_set_runtime_args");
                 if (program_with_callbacks.override_addresses_callback.has_value()) {
                     auto override_addresses_callback = program_with_callbacks.override_addresses_callback.value();
+                    // Deprecated
```

Comment:
Hey Eyon, maybe I'm missing something but why is this API deprecated? We have a template specialization that allows output tensors of type `Tensors&`.